ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/AttestationAggregator.java

```diff
@@ -0,0 +1,143 @@
+                .ifPresent(
+                    selection_proof -> {
+                      UnsignedLong committeeIndex = attester.getCommittee().getIndex();
+                      committeeIndexToAggregatorInformation.put(
+                          committeeIndex,
+                          new AggregatorInformation(selection_proof, attester.getValidatorIndex()));
+                    }));
+  }
+
+  public void processAttestation(Attestation newAttestation) {
+    if (aggregationActive.get()) {
+      // Make sure the new attestation only represents a single attester and that
+      // one of our validators is the validator for attestation's committee index
+      if (isSingleAttester(newAttestation)
+          && committeeIndexToAggregatorInformation
+              .keySet()
+              .contains(newAttestation.getData().getIndex())) {
+
+        Bytes32 attestationDataHashTreeRoot = newAttestation.getData().hash_tree_root();
+        Optional<Attestation> aggregateAttestation =
+            ChainStorage.get(attestationDataHashTreeRoot, dataHashToAggregate);
+
+        // If there exists an old aggregate attestation with the same Attestation Data,
+        // and the new Attestation represents a new attester, add the signature of the
+        // new attestation to the old aggregate attestation.
+        if (aggregateAttestation.isPresent()
+            && representsNewAttesterSingle(aggregateAttestation.get(), newAttestation)) {
+
+          aggregateAttestation(aggregateAttestation.get(), newAttestation);
+
+        }
+
+        // If the attestation message hasn't been seen before:
+        // - add it to the aggregate attestation map to aggregate further when
+        // another attestation with the same message is received
+        // - add it to the list of aggregate attestations for that commiteeeIndex
+        // to broadcast
+        else if (aggregateAttestation.isEmpty()) {
+          ChainStorage.add(attestationDataHashTreeRoot, newAttestation, dataHashToAggregate);
+          UnsignedLong committeeIndex = newAttestation.getData().getIndex();
+          committeeIndexToAggregatesList.computeIfAbsent(committeeIndex, k -> new ArrayList<>());
+          List<Attestation> aggregatesList = committeeIndexToAggregatesList.get(committeeIndex);
+          aggregatesList.add(newAttestation);
+        }
+      }
+    }
+  }
+
+  private void aggregateAttestation(
+      Attestation oldAggregateAttestation, Attestation newAttestation) {
+    // Set the bit of the new attester in the aggregate attestation
+    oldAggregateAttestation
+        .getAggregation_bits()
+        .setBit(getAttesterIndexIntoCommittee(newAttestation));
+
+    // Aggregate signatures
+    List<BLSSignature> signaturesToAggregate = new ArrayList<>();
+    signaturesToAggregate.add(oldAggregateAttestation.getAggregate_signature());
+    signaturesToAggregate.add(newAttestation.getAggregate_signature());
+    oldAggregateAttestation.setAggregate_signature(
```

Comment:
I think there are some thread-safety issues with `Attestation` - the fields are mutable / non-final but not volatile.

Reply:
I refactored this function to solely be for aggregating signatures and made only this synchronized.